docs/architecture/adr-054-crypto-encoding-2.md

```diff
@@ -0,0 +1,85 @@
+# ADR 054: Crypto encoding (part 2)
+
+## Changelog
+
+\*2020-2-27: Created
+
+## Context
+
+Amino has been a pain point of many users in the ecosystem. While Tendermint does not suffer greatly from the performance degradation introduced by amino, we are making an effort in moving the encoding format to a widely adopted format, [Protocol Buffers](https://developers.google.com/protocol-buffers). With this migration a new standard is needed for the encoding of keys. This will cause ecosystem wide breaking changes.
+
+Currently amino encodes keys as `<PrefixBytes> <Length> <ByteArray>`.
+
+> With keys becoming a [oneof type](https://developers.google.com/protocol-buffers/docs/proto3#oneof) we have the option of removing the prefix bytes.
+
+## Decision
+
+Transitioning from a fixed size byte array to bytes would be the first step. This will enable usage of [cosmos-proto](https://github.com/regen-network/cosmos-proto) interface type. This removes boiler plate needed for oneof types.
+
+The approach that will be taken to minimize headaches for users is one where all encoding of keys will shift to protobuf and where amino encoding is relied on, there will be custom marshal and unmarshal functions.
```

Comment:
Maybe out of scope for this document, but do we have a list of places where we'll still need to use amino?

Reply:
Ah, I see the comment on this below 😅

Reply:
i did a dive into the code where we encode keys and how we do it. Currently I have not found any reliance on amino encoded bytes but would like to spend sometime doublechecking as i go through the work

Reply:
I'd also like for us to think about the storage format (currently, amino-encoded objects).

Reply:
this will change the proto encoding. that is why a hardfork is necessary to integrate this into chains, unless you have other ideas on how to handle state encoding?